kw6/reader.py

```diff
@@ -73,44 +79,29 @@ def __iter__(self):
         stream.close()
 
     def __len__(self):
-        try:
-            return self.assumptuous_length()
-        except Exception:
-            self.stream.seek(settings.N_BYTES_VERSION)
-            if self.empty():
-                return 0
-
-            position_header = None
-            while not self.empty():
-                byte_position = self.stream.tell()
-                position_header = Position.skip_(self.stream)
-                self.cached_byte_positions[position_header.frame_index] = byte_position
-
-            return position_header.frame_index - self.initial_frame_index + 1
+        for _ in range(100):
+            assumptuous_length = self.assumptuous_length()
+            try:
+                max_position = self[self.initial_frame_index + assumptuous_length - 1]
+                max_byte_position = self.cached_byte_positions[max_position.header.frame_index]
+                if self.n_bytes == max_byte_position + max_position.header.n_frame_bytes:
+                    return assumptuous_length
+            except Exception:
+                pass
 
-    def assumptuous_length(self):
-        self.stream.seek(settings.N_BYTES_VERSION)
-        position_header = self.position_header()
-        self.stream.seek(0, io.SEEK_END)
-        end_position = self.stream.tell()
-        assumptuous_length = (
-            end_position - settings.N_BYTES_VERSION
-        ) / position_header.n_frame_bytes
+        raise Exception(f"Failed to calculate length of {self.path}")
 
-        if np.abs(assumptuous_length - np.round(assumptuous_length)) <= 1e-9:
-            assumptuous_length = int(assumptuous_length)
-        else:
-            raise Exception(f"Unexpected length {assumptuous_length} of kw6 file")
+    def assumptuous_length(self, from_frame_index=None):
+        if from_frame_index is None:
+            from_frame_index = max(self.cached_byte_positions.keys())
 
-        expected_frame_index = self.initial_frame_index + assumptuous_length - 1
-        self.stream.seek(end_position - position_header.n_frame_bytes)
-        position = Position.from_stream_(self.stream)
-        if position.header.frame_index != expected_frame_index:
-            raise Exception(
-                f"Expected to find {expected_frame_index}, "
-                f"but found {position.header.frame_index} instead."
-            )
-        return assumptuous_length
+        from_position = self[from_frame_index]
+        max_byte_position = self.cached_byte_positions[from_frame_index]
+        return int(
+            (self.n_bytes - max_byte_position) / np.clip(from_position.header.n_frame_bytes, 100, 400000)
```

Comment:
Why is `np.clip` needed here?

Reply:
It's more of an artifact that I forgot about though it's not entirely unreasonable. I'll remove it